Chunk initial topography

anne-glerum opened the pull request:
@gassmoeller 
Hi Rene,
do you think you could give this geometry model a try? I've added a test that you could run. Also, right now, that test fails on my laptop because I put in an AssertThrow for older deal.II versions that try to add non-zero initial topography. What would be the best way to go around this?
I'd like a more compact way to put in the derivative gradient, but that's for tomorrow :)
Thank you!

anne-glerum commented:
@gassmoeller Thanks for testing! I'll have a look today, now that I can test both versions.

gassmoeller commented:
Great, that sounds like you found the problem :+1:. Let us stay consistent and use the signal to work around it. I do not quite remember the logic in `AsciiDataBoundary::get_data_component`, but I remember there were a lot of special cases to consider, and it is complicated enough already.
How did you end up running into the HyperShellBoundary error? Should not the boundary objects only be created and attached if we use deal.II < 9.0, and the initial topography only be used with deal.II >= 9.0?

gassmoeller commented:
/run-tests

anne-glerum commented:
Alright, I'll try and use the signal and address your comments. 

I ran into the HyperShellBoundary error when I enabled initial topography with deal.II < 9.0 (because I knew this used to work).

anne-glerum commented on source/geometry_model/chunk.cc:
```diff
@@ -408,9 +660,21 @@ namespace aspect
+      // depth is therefore always positive
+      const double outer_radius = manifold.get_radius(position);
+      const Point<dim> rtopo_phi_theta = manifold.pull_back_sphere(position);
+      return std::max(0.0, outer_radius - rtopo_phi_theta[0]);
```
Depth here is computed as the difference between the radius of modified surface (given by get_radius) and the modified point (given by pull_back_sphere).

anne-glerum commented:
Not yet ready for another review

naliboff commented:
@gassmoeller @anne-glerum - I was reading through the documentation on the ascii data Initial topography` [plugin](https://github.com/geodynamics/aspect/blob/master/source/geometry_model/initial_topography_model/ascii_data.cc#L151-L152) and it states that the Chunk geometry is currently supported. It looks like the functionality from this PR is not currently added in through another PR. Unless I'm missing something, I'll go ahead and update the documentation.

bangerth commented:
@anne-glerum -- I think we dropped the ball on this. Can you see if you can rebase this on current master?

anne-glerum commented:
@bangerth Yep, definitely dropped the ball on this one. I'll get back to this in a week!

naliboff commented:
@anne-glerum - I rebased this recently and just had one issue or two issues to fix (I think). I’ll take a look over the weekend and let you know if I can get it to compile.

anne-glerum commented:
Hi @gassmoeller @naliboff , I've rebased, updated and simplified this pr and I think it's ready for another review. Repeating the test of which I posted figures above, now results in a flow field from high to low topography:
![chunk_sphere_extremetopo_noblob_v](https://user-images.githubusercontent.com/7631624/55624134-4cba2a00-57a5-11e9-9995-daac0cd3f00e.png)
The tester fails as the test only works for dealii >=9, while the tester uses dealii 8.5. Is there a way to only run a test for a specific dealii version?

anne-glerum commented:
For reference, I'm posting some snapshots of 1 timestep chunk models with varying resolution/refinement/topography using the updates in this pr. Models are based on the test`chunk_topography_postprocessor.prm` (rising blob, no slip BC except for free surface).

No topography, only global refinement (bottom) or with adaptive refinement (top), refinement increases to the right:
![chunk_sphere_notopo_ref_amr](https://user-images.githubusercontent.com/7631624/55625031-e256b900-57a7-11e9-9045-be76f2055e07.png)
![chunk_sphere_notopo_ref_amr_strianrate](https://user-images.githubusercontent.com/7631624/55625074-fbf80080-57a7-11e9-9286-8424f6473bcb.png)
Comparison of chunk (top) and shell (bottom) for adaptive refinement and without topography:
![chunk_shell_sphere_notopo_ref_amr_vel](https://user-images.githubusercontent.com/7631624/55625167-45485000-57a8-11e9-9257-09bceee9af0e.png)
![chunk_shell_sphere_notopo_ref_amr_strainrate](https://user-images.githubusercontent.com/7631624/55625175-4d07f480-57a8-11e9-886c-8e514ebcaa79.png)
Chunk with global refinement with initial topography without (top) and with (bottom) a rising sphere:
![chunk_sphere_topo_noblob_ref_v](https://user-images.githubusercontent.com/7631624/55625260-817bb080-57a8-11e9-991c-f946cfd8d9fe.png)

naliboff commented:
@anne-glerum - Thanks for updating all of this! I compiled the branch to run a few tests, but unfortunately both the 2D chunk and the models I made are failing at this line in` utilities.cc`:
```
      if (this->get_time() - first_data_file_model_time >= 0.0 ||
          (dynamic_cast<const GeometryModel::Chunk<dim>*>(&this->get_geometry_model()) != 0 &&
           dynamic_cast<const InitialTopographyModel::AsciiData<dim>*>(&this->get_initial_topography_model()) != 0 &&
           std::isnan(this->get_time())))
```
Is the test successfully running on your end? If so, perhaps it is a deal.II version issue?

anne-glerum commented:
Hi @naliboff , yes the test works for me. What error message do you get?

naliboff commented:
@anne-glerum The error message is just a floating point exception, but clearly no division by zero in the failing line. Perhaps the initial data file model time may not be initialized yet? Let me know if digging deeper would be helpful!

FYI, I built the branch with deal.II v9.0.1.

```
Thread 1 "aspect" received signal SIGFPE, Arithmetic exception.
0x000055555613760b in aspect::Utilities::AsciiDataBoundary<2>::get_data_component (this=this@entry=0x5555568cdc20, 
    boundary_indicator=<optimized out>, position=..., component=component@entry=0)
    at /data1/Software/aspect/chunk_initial_topo/aspect/source/utilities.cc:2177
2177	      if (this->get_time() - first_data_file_model_time >= 0.0 ||
(gdb) 
```

anne-glerum commented:
@naliboff I've switched the conditional statements. Only in the very specific case of Initial topography specified by an ascii data table, we should get to these lines before time is set to something other than NaN. Now we should hit the condition that checks for NaN first and hopefully that solves the problem. If not, some more info would be helpful and I'll try to reproduce the error.

naliboff commented:
@anne-glerum - hmmm, still getting the same error with the last commit. I'll dig into the debugger tomorrow and see if I can pinpoint a solution on my end.

anne-glerum commented:
new is_time_initialized parameter could be updated after #2529

gassmoeller commented on include/aspect/geometry_model/initial_topography_model/interface.h:
```diff
@@ -73,6 +73,12 @@ namespace aspect
+         * Return the value of the elevation at the given point.
+         */
+//        virtual
+//        Tensor<1,dim-1> vector_gradient (const Point<dim> &p) const = 0;
```
I think this function is still here and unnecessary?

gassmoeller commented on include/aspect/geometry_model/chunk.h:
```diff
@@ -328,9 +394,41 @@ namespace aspect
             double point1_lon;
+            double inner_radius;
+
+            double max_depth;
```
Please add comments for `max_depth` and `inner_radius`

gassmoeller commented on include/aspect/utilities.h:
```diff
@@ -585,6 +585,19 @@ namespace aspect
+        /**
+         * Returns the gradient of the function based on the bilinear
+         * interpolation of the data (velocity, temperature, etc. - according
+         * to the used plugin) in Cartesian coordinates. TOOD Cartesian??
```
Remove TODO

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -32,6 +33,10 @@
 #include <deal.II/grid/tria_boundary_lib.h>
 #endif
 
+#if DEAL_II_VERSION_GTE(9,0,0)
```
No need for DEAL_II_VERSION_GTE(9,0,0) anymore, because we now require 9.0.

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -40,31 +45,46 @@ namespace aspect
+
+      // prior to deal.II 9, we do not apply initial topography
+#if !DEAL_II_VERSION_GTE(9,0,0)
+      const double phi = chart_point[1]; // Longitude
```
This case no longer happens, remove the if and the line.

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -97,15 +117,149 @@ namespace aspect
+    {
+#if !DEAL_II_VERSION_GTE(9,0,0)
+      return push_forward_sphere(r_phi_theta);
+#else
```
Only the `#else` remains.

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -97,15 +117,149 @@ namespace aspect
+    {
+#if !DEAL_II_VERSION_GTE(9,0,0)
+      return pull_back_sphere(x_y_z);
+#else
```
Same as above

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -186,7 +340,6 @@ namespace aspect
     }
 
 
-
```
We usually have 3 empty lines between functions, please keep this line

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -198,19 +351,112 @@ namespace aspect
     void
     Chunk<dim>::initialize ()
     {
+#if !DEAL_II_VERSION_GTE(9,0,0)
```
This case does not happen anymore

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -464,7 +724,6 @@ namespace aspect
     }
 
 
-
```
keep the line

gassmoeller commented on source/utilities.cc:
```diff
@@ -1829,7 +1838,6 @@ namespace aspect
     }
 
 
-
```
keep the line

gassmoeller commented on source/utilities.cc:
```diff
@@ -2165,7 +2173,11 @@ namespace aspect
     {
-      if (this->get_time() - first_data_file_model_time >= 0.0)
+      // For initial ascii data topography, we need access to the data before get_time() is set
+      if ( (dynamic_cast<const GeometryModel::Chunk<dim>*>(&this->get_geometry_model()) != nullptr &&
```
please add a comment why these geometries are special

gassmoeller commented on source/utilities.cc:
```diff
@@ -2200,6 +2212,49 @@ namespace aspect
+          return time_weight * gradients + (1 - time_weight) * old_gradients;
+        }
+      else
+        return Tensor<1,dim-1>();
```
This way you silently return 0.0 for other geometries than the Chunk? Please make this an `AssertThrow(false, ...)` so that nobody accidentally calls this function for other geometries.

anne-glerum replied:
I think return this Tensor makes sense for the other condition (`this->get_time() - first_data_file_model_time >= 0.0`). Then if we're not far enough in time yet, 0 is returned. The same is done for `get_data_component`.

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -198,19 +351,112 @@ namespace aspect
+
+      // Grab lon,lat coordinates
+      Point<dim-1> surface_point;
+      for (unsigned int d=0; d<dim-1; d++)
```
just for convention `++d`

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -186,7 +340,6 @@ namespace aspect
 
 
-
 #if DEAL_II_VERSION_GTE(9,0,0)
```
can remove the ifdef

anne-glerum commented:
@gassmoeller I've addressed the comments.

anne-glerum commented:
This PR has been superseded by #3039, so I'm closing it.
